src/libopenrave-core/environment-core.h

```diff
@@ -2986,181 +2986,179 @@ class Environment : public EnvironmentBase
+        std::unordered_map<string_view, int> existingBodyIndicesById; // Map of body id -> environmentBodyIndex into vBodies
+        std::unordered_map<string_view, int> existingBodyIndicesByName; // Map of body name -> environmentBodyIndex into vBodies
+        for (size_t bodyIndex = 0; bodyIndex < vBodies.size(); bodyIndex++) {
+            // It's possible for holes to exist in vBodies since when bodies are removed from the environment their entries in _vecbodies are just nulled out.
+            if (!vBodies[bodyIndex]) {
+                continue;
             }
+
+            const KinBody& body = *vBodies[bodyIndex];
+            if (bodyIdsToUpdate.find(body.GetId()) == bodyIdsToUpdate.end() && bodyNamesToUpdate.find(body.GetName()) == bodyNamesToUpdate.end()) {
+                continue;
+            }
+            existingBodyIndicesById[body.GetId()] = bodyIndex;
+            existingBodyIndicesByName[body.GetName()] = bodyIndex;
         }
 
-        // Set of indices that have already been used for vBodies
-        std::unordered_set<int> usedBodyIndexSet;
+        // Keep a list of the bodies that we have matched to infos - in the event that the caller did _not_ specify OnlySpecifiedBodiesExact,
+        // when we are finished we will want to remove all bodies in the environment that are _not_ in this list.
+        // Store raw pointers here instead of shared pointers because we don't need to track ownership, just membership.
+        std::unordered_set<KinBody*> setBodiesWithMatchingInfos;
 
         // internally manipulates _vecbodies using _AddKinBody/_AddRobot/_RemoveKinBodyFromIterator
-        for(int inputBodyIndex = 0; inputBodyIndex < (int)info._vBodyInfos.size(); ++inputBodyIndex) {
-            const KinBody::KinBodyInfoConstPtr& pKinBodyInfo = info._vBodyInfos[inputBodyIndex];
+        for(const KinBody::KinBodyInfoPtr& pKinBodyInfo : info._vBodyInfos ) {
             const KinBody::KinBodyInfo& kinBodyInfo = *pKinBodyInfo;
             RAVELOG_VERBOSE_FORMAT("env=%s, id '%s', name '%s', _vGrabbedInfos=%d", GetNameId()%pKinBodyInfo->_id%pKinBodyInfo->_name%pKinBodyInfo->_vGrabbedInfos.size());
             RobotBase::RobotBaseInfoConstPtr pRobotBaseInfo = OPENRAVE_DYNAMIC_POINTER_CAST<const RobotBase::RobotBaseInfo>(pKinBodyInfo);
-            KinBodyPtr pMatchExistingBody; // matches to pKinBodyInfo
-            int bodyIndex = -1; // index to vBodies to use. -1 if not used
-            {
-                // find existing body in the env
-                std::vector<KinBodyPtr>::iterator itExistingSameId = vBodies.end();
-                std::vector<KinBodyPtr>::iterator itExistingSameName = vBodies.end();
-                std::vector<KinBodyPtr>::iterator itExistingSameIdName = vBodies.end();
-
-                if( updateMode == UFIM_OnlySpecifiedBodiesExact ) {
```

Comment:
Is `UFIM_OnlySpecifiedBodiesExact` correctly handled in the new version?

Reply:
Can you elaborate? The `UFIM_OnlySpecifiedBodiesExact` case should be handled down on line 3301. See testopenrave PR 141 for the relevant tests.

Reply:
I think the old logic here was to ignrore duplicated matching in case of `UFIM_OnlySpecifiedBodiesExact`, i.e. if the caller passed in a list of kinbody info that has duplicates. I think that's automatically an error, which is guarded by exceptions added in the beginning of the function.